### DISCUSSION TITLE
How to maintain global pool of db connection and use it in each and every request?

### DISCUSSION THREAD

**Discussion by gauravsaini964** [fastapi/fastapi#9097] [Questions] [answered]
Hi,
I am using asyncpg for connecting to postgres database. I am not able to figure out how to establish a connection on app boot and maintain pool which I can use throughout the app.

I have the following architecture:
Each request goes through middleware, here it performs a local key check and verifies the user against firebase.
After that, request goes to the handler, here I need a DB connection from the central pool. Almost every request uses a DB connection.

**Comment by old-syniex** [comment]
Store it on app.state, and access it with request.app.state

**Comment by gauravsaini964** [comment]
> Store it on app.state, and access it with request.app.state

Is this approach appropriate? If not, how should I refactor this according to your approach?

db.py
```
import asyncpg

# from psycopg2 import pool
from configs.settings import settings


class Database:
    def __init__(self):
        self.user = settings.POSTGRES_USER
        self.password = settings.POSTGRES_PASSWORD
        self.host = settings.POSTGRES_SERVER
        self.port = "5432"
        self.database = settings.POSTGRES_DB
        self._cursor = None

        self._connection_pool = None
        self.con = None

    async def connect(self):
        if not self._connection_pool:
            try:
                self._connection_pool = await asyncpg.create_pool(
                    min_size=1,
                    max_size=10,
                    command_timeout=60,
                    host=self.host,
                    port=self.port,
                    user=self.user,
                    password=self.password,
                    database=self.database,
                )

            except Exception as e:
                print(e)

    async def fetch_rows(self, query: str):
        if not self._connection_pool:
            await self.connect()
        else:
            self.con = await self._connection_pool.acquire()
            try:
                result = await self.con.fetch(query)
                return result
            except Exception as e:
                print(e)
            finally:
                await self._connection_pool.release(self.con)



```

db_session.py

```
from .db_connection import Database

database_instance = Database()


```

main.py
```
from fastapi import FastAPI
from fastapi.middleware.cors import CORSMiddleware


from routes import items, user

from utils.middleware import middleware
from configs import open_api_tags
from configs.settings import settings
from db.db_session import database_instance

app = FastAPI(
    title=settings.PROJECT_NAME,
    description=settings.PROJECT_DESCRIPTION,
    version="0.0.1",
    openapi_tags=open_api_tags.tags_metadata,
)


app.add_middleware(
    CORSMiddleware,
    allow_origins=settings.BACKEND_CORS_ORIGINS,
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)


app.middleware("http")(middleware)


@app.on_event("startup")
async def startup():
    await database_instance.connect()


app.include_router(user.router)
app.include_router(items.router, prefix="/items")

```

and in route.py

```
from fastapi import APIRouter
from pydantic import BaseModel


from db.db_session import database_instance


router = APIRouter()


class UserCreateQuery(BaseModel):
    idtoken: str


@router.post("/auth/create/")
async def UserCreateAPI(user: UserCreateQuery):
    database_instance.fetch_rows("SELECT 1")
    return user


```

**Comment by old-syniex** [comment, marked as answer]
main.py
```python
from fastapi import FastAPI
from fastapi.middleware.cors import CORSMiddleware


from routes import items, user

from utils.middleware import middleware
from configs import open_api_tags
from configs.settings import settings
from db.db_session import database_instance

app = FastAPI(
    title=settings.PROJECT_NAME,
    description=settings.PROJECT_DESCRIPTION,
    version="0.0.1",
    openapi_tags=open_api_tags.tags_metadata,
)


app.add_middleware(
    CORSMiddleware,
    allow_origins=settings.BACKEND_CORS_ORIGINS,
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)


app.middleware("http")(middleware)


@app.on_event("startup")
async def startup():
    await database_instance.connect()
    app.state.db = database_instance


app.include_router(user.router)
app.include_router(items.router, prefix="/items") 
```

route.py
```python
from fastapi import APIRouter,Request
from pydantic import BaseModel


from db.db_session import database_instance


router = APIRouter()


class UserCreateQuery(BaseModel):
    idtoken: str


@router.post("/auth/create/")
async def UserCreateAPI(request: Request, user: UserCreateQuery):
    request.app.state.db.fetch_rows("SELECT 1")
    return user
```

that should work,
please check and let me know

**UPD**: Fixed `on_startup` event thanks to @frutik (https://github.com/fastapi/fastapi/discussions/9097#discussioncomment-5156902)

**Comment by old-syniex** [comment]
@gauravsaini964 can you please let me know if this solved your problem?

**Comment by frutik** [comment]
@Syniex 

in my case

`AttributeError: 'coroutine' object has no attribute 'fetch_rows'`

**Comment by frutik** [comment]
Seems 

```
@app.on_event("startup")
async def startup():
    await database_instance.connect()
    app.state.db = database_instance
```

Solves the issue

**Comment by old-syniex** [comment]
My bad I didn't see your connect function on Database is async.

Glad it worked.

**Comment by houmie** [comment]
According to the FastAPI documentation, for the async approach, it is recommended to use `Databases` and SqlAlchemy core.   How do I create the pool as shown here ```await asyncpg.create_pool()``` and inject it into ```db = databases.Database(settings.SQLALCHEMY_DATABASE_URI)``` ?

**Comment by healiseu** [comment]
I have tried to use a similar approach, but it didn't work when I tried to send multiple (e.g. 100) requests concurrently. I was getting an error (**asyncpg.exceptions._base.InterfaceError cannot perform operation: another operation is in progress**). My mistake was that in the middleware I was trying to acquire a database connection from the pool and then release it.

After researching several posts on this issue and how to utilize asyncpg pooling on server applications, that handle frequent requests and need the database connection for a short period time while handling a request I found this stackoverflow post, [how to do persistent database connection in fastapi][1] which solved my problem.

The key difference that this post makes is that you don't want to share a connection but the connections pool. Therefore you add the pool to a request.state in the middleware and you pass the pool to every route that is executing asyncpg database queries. This is also the method they used in [FastAPI framework benchmarks][2].

As a bonus point, consider also that you should set asyncpg pool object parameter **max_inactive_connection_lifetime** accordingly. The default value is 300sec. Here is the relevant github issue MagicStack/asyncpg#309


[1]: https://stackoverflow.com/questions/63270196/how-to-do-persistent-database-connection-in-fastapi
[2]: https://github.com/TechEmpower/FrameworkBenchmarks/blob/master/frameworks/Python/fastapi/app_orjson.py

**Comment by MarkParker5** [comment]
If you are using sqlalchemy and getting error: "asyncpg.exceptions._base.InterfaceError: cannot perform operation: another operation is in progress", [try set poolclass=NullPool in create_async_engine](https://stackoverflow.com/a/73485900/11677524)

**Comment by bvsn** [comment]
@MarkParker5 yeah, it's a good solution but there is became no sense in async pool managing. In a such case, you should use pgpool or pgbouncer to reuse your connections.

During a lot of experiments I've made a couple of conclusions:
1. If you want to use `asyncpg` and have fully async querying, use `Databases` with a shared connection through the `app.state`. But there is weak support of SQLModel, so you should use a core SQLAlchemy.
2. If you want a typing and comfortable querying, use SQLModel with the NullPool and `pgpool` or `pgbouncer`. But forget about controlling your connections in your app (Do you really need it?).

So, my choice is to use asyncpg with the AsyncSession from SQLModel and NullPool setup for an engine. Hope, in the future I can switch to QueuePool and avoid async managing issues.

**Update**. If you're keeping a session only on a transaction level, you can use a `dispose` of connection at the end of the transaction. Look at these examples 👉 https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html

**Comment by heskarioth** [comment]
> @MarkParker5 yeah, it's a good solution but there is became no sense in async pool managing. In a such case, you should use pgpool or pgbouncer to reuse your connections.
> 
> During a lot of experiments I've made a couple of conclusions:
> 
> 1. If you want to use `asyncpg` and have fully async querying, use `Databases` with a shared connection through the `app.state`. But there is weak support of SQLModel, so you should use a core SQLAlchemy.
> 2. If you want a typing and comfortable querying, use SQLModel with the NullPool and `pgpool` or `pgbouncer`. But forget about controlling your connections in your app (Do you really need it?).
> 
> So, my choice is to use asyncpg with the AsyncSession from SQLModel and NullPool setup for an engine. Hope, in the future I can switch to QueuePool and avoid async managing issues.
> 
> **Update**. If you're keeping a session only on a transaction level, you can use a `dispose` of connection at the end of the transaction. Look at these examples 👉 https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html

Could you please show example for second suggestion?

**Comment by bvsn** [comment]
@heskarioth it may be something like this:

```
from sqlalchemy.engine.url import URL
from sqlalchemy.exc import SQLAlchemyError
from sqlalchemy.ext.asyncio import create_async_engine
from sqlalchemy.orm import sessionmaker
from sqlalchemy.pool import NullPool
from sqlmodel.ext.asyncio.session import AsyncSession

ASYNC_DATABASE_URL = URL("postgresql+asyncpg", username=username, password=password, host=host, port=port, database=database)

async_engine = create_async_engine(ASYNC_DATABASE_URL, future=True, echo=False, poolclass=NullPool)
async_session = sessionmaker(bind=async_engine, autocommit=False, autoflush=False, expire_on_commit=False, class_=AsyncSession)

async def run(query):
    with async_session() as session:
        try:
            result = await session.exec(statement=query)
        except SQLAlchemyError as sql_ex:
            await session.rollback()
            raise sql_ex
        finally:
            await session.commit()

    return result
```

Then you will have to deal with different approaches specific to your application, but in general, it may work like that.
BTW. You could remove NullPool, it may work. Not tested it by myself 🤷‍♂️ You mat test it with pytest for example.

**Comment by anandhu-gopi** [comment]
**Is it possible to use a class-based approach to maintain a global pool of database connections and utilize the Depends functionality for proper typing and editor hints?**


The DataBasePool class handles the creation, retrieval, and teardown of the database connection pool. The startup and shutdown event handlers in main.py ensure that the connection pool is appropriately set up and closed when the application starts and stops.



**database.py**
```python

import asyncpg
from asyncpg import Pool
from typing import Optional


class UninitializedDatabasePoolError(Exception):
    def __init__(
        self,
        message="The database connection pool has not been properly initialized.Please ensure setup is called",
    ):
        self.message = message
        super().__init__(self.message)


class DataBasePool:

    _db_pool: Optional[Pool] = None

    @classmethod
    async def setup(cls, timeout: Optional[float] = None):
        cls._db_pool = await asyncpg.create_pool(database="postgres", user="postgres")
        cls._timeout = timeout

    @classmethod
    async def get_pool(cls):
        if not cls._db_pool:
            raise UninitializedDatabasePoolError()
        return cls._db_pool

    @classmethod
    async def teardown(cls):
        if not cls._db_pool:
            raise UninitializedDatabasePoolError()
        await cls._db_pool.close()

```
**main.py**
```python

from typing import Optional
import asyncpg
from fastapi import Depends, FastAPI

from database import DataBasePool

app = FastAPI()


@app.on_event("startup")
async def startup():
    """Initialize the application server."""
    # Create a database connection pool
    await DataBasePool.setup()


@app.on_event("shutdown")
async def shutdown():
    # close the database connection pool on shutdown
    await DataBasePool.teardown()


@app.get("/")
async def handle(db_pool:asyncpg.Pool = Depends(DataBasePool.get_pool)):
    """Handle incoming requests."""
    async with db_pool.acquire() as connection:
        async with connection.transaction():
            # do something with the db_pool
            result = await connection.fetchval("SELECT COUNT(*) FROM users")
            return {"result": result}


if __name__ == "__main__":
    import uvicorn
    uvicorn.run(app, host="0.0.0.0", port=8000)
```

P.S: Saw this recommendation of using a connection pool in the asynpg documentation https://magicstack.github.io/asyncpg/current/usage.html#connection-pools

**Comment by YuriFontella** [comment]
Is it still a good way to store the connection in `app.state` and then use it with `request.app.state`?
It doesn't seem very appropriate to always have to call (request: Request) on my routes

**Comment by WolfEYc** [comment]
ngl yall make it more complicated then it has to be...

postgres_connection.py
```py
async def create_pool() -> asyncpg.Pool:
    pool = await asyncpg.create_pool(DB_URL)
    if pool is None:
        raise ValueError("Could not create connection pool")
    return pool

pgpool = await create_pool()
 ```
 main.py
 ```py
 @asynccontextmanager
async def lifespan():
    yield
    await pgpool.close()
```

as much as i love
```py
async def handle(db_pool:asyncpg.Pool = Depends(DataBasePool.get_pool)):
    """Handle incoming requests."""
    async with db_pool.acquire() as connection:
        async with connection.transaction():
            # do something with the db_pool
            result = await connection.fetchval("SELECT COUNT(*) FROM users")
```
i think we all would have a better time just

some_file_literally_anywhere_in_the_codebase.py
```py
async def gotta_do_some_db_calls():
    await pgpool.execute("DROP DATABASE prod")
```
